### PR TITLE
Add org.thonny.Thonny

### PR DESCRIPTION
**Pull request by aivarannamaa**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project.
- [x] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

### CODE REVIEW

**Comment by jwillikers** [comment]
bot, build org.thonny.Thonny

**Comment by jwillikers** [comment]
@aivarannamaa You might need to cut a new release with the those fixes to the AppData so that they are validated properly. Just be sure to add a corresponding release section to the AppData file, too. Then all you need to do is update the corresponding tag and commit hash in the `org.thonny.Thonny.yaml` manifest in this PR. I've got a snippet of that section below.

```
   sources:
      - type: git
        url: https://github.com/thonny/thonny.git
        tag: v3.3.13
        commit: b46e5f60f36c200144f450f919b73256629511d7
```

File: org.thonny.Thonny.yaml [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+      - /lib/python3.8/test
+      - /lib/python3.8/turtle*
+      - /lib/python3.8/__pycache__/turtle.*
+      - /lib/python3.8/*/test
+      - /lib/python3.8/*/tests
+    cleanup-commands:
+      - find "/lib" -name "lib*.a" -exec rm {} ";"
+    sources:
+      - type: archive
+        url: https://www.python.org/ftp/python/3.8.8/Python-3.8.8.tar.xz
```
Comment (hfiguiere) [inline org.thonny.Thonny.yaml:75]: do you really need to build python?

File: python3-modules.json [hunk 2]
```diff
@@ -0,0 +1,589 @@
...
+            "sha256": "806143ae5bfb6a3c6e736a764057db0e6a0e05e338b5630894a5f779cabb4f9b"
+        },
+        {
+            "type": "file",
+            "url": "https://files.pythonhosted.org/packages/87/89/479dc97e18549e21354893e4ee4ef36db1d237534982482c3681ee6e7b57/mccabe-0.6.1-py2.py3-none-any.whl",
+            "sha256": "ab8a6258860da4b6677da4bd2fe5dc2c659cff31b3ee4f7f5d64e79735b80d42"
+        },
+        {
+            "type": "file",
+            "url": "https://files.pythonhosted.org/packages/87/c4/da8ed7f77c85347c0bc50911d27faa563b85b46d136f8ce6c0f8d05f610d/lazy_object_proxy-1.6.0-cp38-cp38-manylinux1_x86_64.whl",
```
Comment (hfiguiere) [inline python3-modules.json:40]: this probably explain the build failure on aarch64.

as well as a few other wheel below that are x86_64 only.

**Comment by jwillikers** [comment]
Looks like mypy might be blocker at the moment on aarch64 according the build log and python/mypy#9688.
We can just skip building for aarch64 until that issue is resolved upstream.
Adding a `flathub.json` file to the PR with the following section will skip building for aarch64.

```json
{
  "skip-arches": ["aarch64"]
}
```

**Review by hfiguiere — CHANGES_REQUESTED** [review]

**Comment by hfiguiere** [comment]
> Looks like mypy might be blocker at the moment on aarch64 according the build log and [python/mypy#9688](https://github.com/python/mypy/issues/9688).
> We can just skip building for aarch64 until that issue is resolved upstream.
> Adding a `flathub.json` file to the PR with the following section will skip building for aarch64.
> 
> ```json
> {
>   "skip-arches": ["aarch64"]
> }
> ```

Just build from source.

The flatpak-builder-tool should already pick source packages.

File: org.thonny.Thonny.yaml (hunk 1 above)
Comment (jwillikers) [reply org.thonny.Thonny.yaml:75]: Is there a better way to include the `tkinter` module bundled with Python? The module is deleted in the org.freedesktop.Sdk.

File: python3-modules.json (hunk 2 above)
Comment (hfiguiere) [reply python3-modules.json:40]: Seems like there is an open issue:
https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/issues/976

File: org.thonny.Thonny.yaml (hunk 1 above)
Comment (hfiguiere) [reply org.thonny.Thonny.yaml:75]: Seems like there is an open issue:
https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/issues/976

**Comment by jwillikers** [comment]
@hfiguiere That's a much better idea. I ended up using the `flatpak-pip-generator` from [this open PR](https://github.com/flatpak/flatpak-builder-tools/pull/170) because the actual one was having problemsi. I'll see what I can do. Thanks for the help!

File: org.thonny.Thonny.yaml (hunk 1 above)
Comment (hfiguiere) [reply org.thonny.Thonny.yaml:75]: and there is a solution there: https://github.com/flathub/flathub/pull/2146/commits/5f6f8d8b63fc623c8eda6b221fd6e40851e9259f

Comment (hfiguiere) [reply org.thonny.Thonny.yaml:75]: or tl;dr https://github.com/iwalton3/tkinter-standalone

**Comment by jwillikers** [comment]
The `flatpak-pip-generator` chokes on the `outdated` package because it's newest version doesn't have the corresponding source on PyPI. I've filed issue outdated/issues/#10 upstream to get this resolved.

File: org.thonny.Thonny.yaml (hunk 1 above)
Comment (jwillikers) [reply org.thonny.Thonny.yaml:75]: Awesome! That helps a lot.

**Comment by jwillikers** [comment]
For now, the simplest way to get things working appears to be to just bundle the minimal number of dependencies.
A lot of the other dependencies require installing various build dependencies and non-Python dependencies, which is quite time consuming.

@aivarannamaa I've updated my branch with the latest changes [here](https://github.com/jwillikers/flathub/tree/org.thonny.Thonny).

**Comment by jwillikers** [comment]
@aivarannamaa I've managed to re-bundle each of the regular and XXL packages in my [fork](https://github.com/jwillikers/flathub/tree/org.thonny.Thonny).

**Comment by aivarannamaa** [comment]
bot, build org.thonny.Thonny

**Review by jwillikers — APPROVED** [review]

**Comment by jwillikers** [comment]
@hfiguiere PTAL. I believe we've addressed all of your concerns, thanks!

**Comment by jwillikers** [comment]
@aivarannamaa Do you want to try marking requests for changes resolved / re-request review? That might get things moving again.

File: README.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
+# Thonny Flatpak
+
+This is the official [Thonny](https://thonny.org/) Flatpak.
+
+## Users
+
+This section covers installation instructions and additional information for Thonny users.
```
Comment (hfiguiere) [inline README.md:7]: no user will see that, so that you know.

File: org.thonny.Thonny.yaml [hunk 4]
```diff
@@ -0,0 +1,314 @@
+---
+id: org.thonny.Thonny
+branch: stable
+runtime: org.freedesktop.Sdk
+runtime-version: "20.08"
+sdk: org.freedesktop.Sdk
+command: thonny
+cleanup:
+  - /lib/debug
```
Comment (hfiguiere) [inline org.thonny.Thonny.yaml:9]:
```suggestion
```

Debug info is split out in it's own package, don't remove it.

File: org.thonny.Thonny.yaml [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+        commit: ba946536054f9d27a08aafde21aa18330ce05729
+    modules:
+      - name: tcl
+        buildsystem: autotools
+        subdir: unix
+        build-options:
+          no-debuginfo: true
+        cleanup:
+          - /bin
+          - /lib/debug
```
Comment (hfiguiere) [inline org.thonny.Thonny.yaml:37]:
```suggestion
```

File: org.thonny.Thonny.yaml [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+            url: https://prdownloads.sourceforge.net/tcl/tcl8.6.11-src.tar.gz
+            sha256: 8c0486668586672c5693d7d95817cb05a18c5ecca2f40e2836b9578064088258
+      - name: tk
+        buildsystem: autotools
+        subdir: unix
+        build-options:
+          no-debuginfo: true
+        cleanup:
+          - /bin
+          - /lib/debug
```
Comment (hfiguiere) [inline org.thonny.Thonny.yaml:51]:
```suggestion
```

File: org.thonny.Thonny.yaml [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+          - /man
+        sources:
+          - type: archive
+            url: https://prdownloads.sourceforge.net/tcl/tcl8.6.11-src.tar.gz
+            sha256: 8c0486668586672c5693d7d95817cb05a18c5ecca2f40e2836b9578064088258
+      - name: tk
+        buildsystem: autotools
+        subdir: unix
+        build-options:
+          no-debuginfo: true
```
Comment (hfiguiere) [inline org.thonny.Thonny.yaml:48]:
```suggestion
```
Is this necessary? It's better to have the debug info, they are extracted in a separate package.

File: org.thonny.Thonny.yaml [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+        sha256: 2d475327684562c3a96cc71adf7dc8c4f0565175cf86b6d7a404ff4c771f15f0
+      - type: file
+        url: https://files.pythonhosted.org/packages/cf/5a/25aeb636baeceab15c8e57e66b8aa930c011ec1c035f284170cacb05025e/PyNaCl-1.4.0.tar.gz
+        sha256: 54e9a2c849c742006516ad56a88f5c74bf2ce92c9f67435187c3c5953b346505
+      - type: file
+        url: https://files.pythonhosted.org/packages/9b/77/461087a514d2e8ece1c975d8216bc03f7048e6090c5166bc34115afdaa53/cryptography-3.4.7.tar.gz
+        sha256: 3d10de8116d25649631977cb37da6cbdd2d6fa0e0281d014a5b7d337255ca713
+      - type: file
+        url: https://files.pythonhosted.org/packages/2e/92/87bb61538d7e60da8a7ec247dc048f7671afe17016cd0008b3b710012804/cffi-1.14.6.tar.gz
+        sha256: c9a875ce9d7fe32887784274dd533c57909b7b1dcadcc128a2ac21331a9765dd
```
Comment (hfiguiere) [inline org.thonny.Thonny.yaml:260]: these packages are already install by a previous module. (albeit one is in an older version)

and probable some other I miss.

**Review by hfiguiere — CHANGES_REQUESTED** [review]

File: README.md (hunk 3 above)
Comment (jwillikers) [reply README.md:7]: Hopefully you're right, but I imagine it's possible.

**Comment by jwillikers** [comment]
@aivarannamaa I've got those fixes available in [my branch](https://github.com/jwillikers/flathub/tree/org.thonny.Thonny).

File: org.thonny.Thonny.yaml (hunk 4 above)
Comment (jwillikers) [reply org.thonny.Thonny.yaml:9]: My bad, I was working around flatpak/flatpak#469 but I can workaround this in a better way.

File: org.thonny.Thonny.yaml (hunk 7 above)
Comment (jwillikers) [reply org.thonny.Thonny.yaml:48]: Nope, it's not necessary. I'll remove that option.

File: org.thonny.Thonny.yaml (hunk 8 above)
Comment (jwillikers) [reply org.thonny.Thonny.yaml:260]: I combed through it and removed the duplicate packages.

**Comment by aivarannamaa** [comment]
bot, build org.thonny.Thonny

**Comment by jwillikers** [comment]
Looks good. @aivarannamaa You should be able to re-request review in GitHub.

**Review by jwillikers — APPROVED** [review]

**Comment by nedrichards** [comment]
Hang, on, sorry - is there a reason this is submoduling a fork of flatpak-builder-tools? (since that exposed a bug/error with the github merge action!)

**Comment by jwillikers** [comment]
@nedrichards Not any longer, we use the legitimate `flatpak-builder-tools` now. I had used the fork previously in attempt to get things working, but I figured out the Python packaging. Sorry about that.

**Comment by nedrichards** [comment]
Oh, right, I see now - it's for https://github.com/flatpak/flatpak-builder-tools/pull/170 which is a bit stalled.

**Comment by jwillikers** [comment]
@aivarannamaa I've went ahead and updated my branch [here](https://github.com/jwillikers/flathub/tree/org.thonny.Thonny) to use the actual Flatpak Builder Tools repository.

**Comment by aivarannamaa** [comment]
@jwillikers, should I include the changes you made to `x-checker-data`?

**Comment by jwillikers** [comment]
@aivarannamaa I'd recommend doing so as the Git x-checker-data doesn't take timestamps into account but the JSON x-checker-data does, see flathub/flatpak-external-data-checker#154.

**Comment by aivarannamaa** [comment]
bot, build org.thonny.Thonny

**Review by jwillikers — APPROVED** [review]

File: .gitmodules [hunk 9]
```diff
@@ -0,0 +1,4 @@
+[submodule "flatpak-builder-tools"]
+	path = flatpak-builder-tools
+	url = https://github.com/nanonyme/flatpak-builder-tools.git
+	branch = support-build-isolation
```
Comment (jwillikers) [inline .gitmodules:4]: These two lines should be updated to reflect the true `flatpak-builder-tools` repository:

```
url = https://github.com/flatpak/flatpak-builder-tools.git
branch = master
```

**Review by jwillikers — CHANGES_REQUESTED** [review]

**Comment by aivarannamaa** [comment]
bot, build org.thonny.Thonny

**Review by jwillikers — APPROVED** [review]

**Comment by jwillikers** [comment]
@nedrichards The flatpak-builder-tools submodule has been updated to use the actual repository now.

File: org.thonny.Thonny.yaml [hunk 10]
```diff
@@ -0,0 +1,300 @@
+---
+id: org.thonny.Thonny
+branch: stable
+runtime: org.freedesktop.Sdk
+runtime-version: "20.08"
```
Comment (hfiguiere) [inline org.thonny.Thonny.yaml:5]:
```suggestion
runtime-version: "21.08"
```

21.08 is out and I think it's a good opportunity to use it

File: org.thonny.Thonny.yaml [hunk 11]
```diff
@@ -0,0 +1,300 @@
+---
+id: org.thonny.Thonny
+branch: stable
```
Comment (hfiguiere) [inline org.thonny.Thonny.yaml:3]:
```suggestion
```

This is not needed, flathub use the git branch for that.

**Review by hfiguiere — CHANGES_REQUESTED** [review]

**Comment by aivarannamaa** [comment]
bot, build org.thonny.Thonny

**Comment by aivarannamaa** [comment]
@hfiguiere , after bumping the runtime version, the build bot ends up with this:

```
Processing thonny-3.3.14-py3.9.egg
creating /app/lib/python3.9/site-packages/thonny-3.3.14-py3.9.egg
Extracting thonny-3.3.14-py3.9.egg to /app/lib/python3.9/site-packages
Adding thonny 3.3.14 to easy-install.pth file
error: [Errno 30] Read-only file system: '/app/lib/python3.9/site-packages/easy-install.pth'
```

Do you suggest to simply revert the change or tweak something else?

**Comment by jwillikers** [comment]
@aivarannamaa That's due to me hard-coding the path to be writable for Python 3.8 but the new SDK uses Python 3.9. The troublesome lines are:

```yaml
ensure-writable:
      - /lib/python3.8/site-packages/easy-install.pth
```

That just needs to be Python 3.9 now. I've gone ahead and made this change and updated our dependencies by regenerating them using the newest version of the Freedesktop SDK. As usual, the changes can be found in my branch [here](https://github.com/jwillikers/flathub/tree/org.thonny.Thonny).

File: org.thonny.Thonny.yaml [hunk 12]
```diff
@@ -0,0 +1,299 @@
...
+        --prefix=${FLATPAK_DEST} "websockets==9.1.*" --no-build-isolation
+    sources:
+      - type: file
+        url: https://files.pythonhosted.org/packages/0d/bd/5262054455ab2067e51de331bfbc53a1dfa9071af7c424cf7c0793c4349a/websockets-9.1.tar.gz
+        sha256: 276d2339ebf0df4f45df453923ebd2270b87900eda5dfd4a6b0cfa15f82111c3
+
+  - name: thonny
+    buildsystem: simple
+    ensure-writable:
+      - /lib/python3.8/site-packages/easy-install.pth
```
Comment (hfiguiere) [inline org.thonny.Thonny.yaml:281]:
```suggestion
      - /lib/python3.9/site-packages/easy-install.pth
```

**Comment by aivarannamaa** [comment]
bot, build org.thonny.Thonny

**Review by jwillikers — APPROVED** [review]

**Comment by barthalion** [comment]
@aivarannamaa Please remove `.gitmodules` file – it confuses our GitHub action used for adding new apps.

**Comment by aivarannamaa** [comment]
@jwillikers, would removing `.gitmodules` require additional changes elsewhere?

**Comment by jwillikers** [comment]
@aivarannamaa No, but you might just add a command to the README to download the `flatpak-pip-generator` in the _Update_ section like shown here.

```
wget -L https://raw.githubusercontent.com/flatpak/flatpak-builder-tools/master/pip/flatpak-pip-generator
wget -L https://raw.githubusercontent.com/thonny/thonny/master/packaging/requirements-regular-bundle.txt
python3 flatpak-builder-tools/pip/flatpak-pip-generator --runtime org.freedesktop.Sdk//21.08 -r requirements-regular-bundle.txt
```

 An `rm -rf .gitmodules flatpak-builder-tools` should take care of everything else.

**Comment by aivarannamaa** [comment]
bot, build org.thonny.Thonny

**Review by jwillikers — APPROVED** [review]

**Comment by jwillikers** [comment]
> Please remove `.gitmodules` file – it confuses our GitHub action used for adding new apps.

@barthalion Done.

**Comment by aivarannamaa** [comment]
Anyone who knows, is it possible to give write access to the new repository to @jwillikers ?

**Comment by jwillikers** [comment]
@aivarannamaa I've got them. They added my handle to the `/merge` command above to invite me with write permissions. I'll be keeping an eye on things there. Let me know if you need anything!

**Comment by aivarannamaa** [comment]
@jwillikers, thanks a lot for your help! I now added `flatpak install org.thonny.Thonny` instruction to Thonny's homepage.

**Pull request closed without merging**